\documentclass{article}
\usepackage{fullpage}
\usepackage{hyperref}
\usepackage{authblk}
\usepackage[utf8]{inputenc}

\title{Using vulcan, a package that combines networks and }
\author[1]{Federico M. Giorgi}
\author[1]{Andrew N. Holding}
\affil[1]{Cancer Research UK, Cambridge Institute, Robinson Way, Cambridge United Kingdom}
\date{\today}

\begin{document}
\SweaveOpts{concordance=TRUE}
\maketitle

%-----------
\section{Overview of VULCAN}\label{sec:overview}
Usage of gene regulatory networks to analyze biological systems has witnessed an exponential increase in the last decade, due to the ease of obtaining genome-wide expression data (Basso et al., 2005; Giorgi et al., 2013; Castro et al., 2016). Recently, the VIPER approach to interrogate these network models has been proposed to infer transcription factor activity using the expression of a collection of their putative targets, i.e. their regulon (Alvarez et al., 2016). In the VIPER algorithm, gene-level differential expression signatures are obtained for either individual samples (relative to the mean of the dataset) or between groups of samples, and regulons are tested for enrichment. Ideally, TF regulons can be tested on promoter-level differential binding signatures generated from ChIP-Seq experiments, in order to ascertain the global change in promoter occupancy for gene sets. In our study, we propose an extension of the VIPER algorithm to specifically analyze TF occupancy in ChIP-Seq experiments. Our algorithm, dubbed “VirtUaL ChIP-Seq Analysis through Networks” (VULCAN), uses ChIP-Seq data obtained for a given TF to provide candidate coregulators of the response to a given stimulus (Figure 2). The analysis is based on identifying differentially bound genes and testing their enrichment in the regulon of potential co-regulatory factors.

%-----------
\begin{thebibliography}{99}
\bibitem{Alvarez2013} Alvarez MJ, Shen Y, Giorgi FM, Lachmann A, Ding BB, Ye BH \& Califano, A. Functional characterization of somatic mutations in cancer using network-based inference of protein activity. Nature Genetics (In press) (2016).
\end{thebibliography}
%----------
\end{document}
